<a href="https://colab.research.google.com/github/pockypoem/orbit-ai/blob/main/Ensemble_Technique_Brain_Stroke.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Deskripsi

Ini adalah file kode dalam rangka pembelajaran mengenai Ensemble Technique pada AI Domain Data Science pada dataset Brain Stroke

## EDA

In [27]:
import pandas as pd
import numpy as np

from imblearn.over_sampling import SMOTE

In [38]:
data = pd.read_csv('brain_stroke.csv')
data.head()

gender   age  hypertension  heart_disease ever_married      work_type  \
0    Male  67.0             0              1          Yes        Private   
1    Male  80.0             0              1          Yes        Private   
2  Female  49.0             0              0          Yes        Private   
3  Female  79.0             1              0          Yes  Self-employed   
4    Male  81.0             0              0          Yes        Private   

  Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0          Urban             228.69  36.6  formerly smoked       1  
1          Rural             105.92  32.5     never smoked       1  
2          Urban             171.23  34.4           smokes       1  
3          Rural             174.12  24.0     never smoked       1  
4          Urban             186.21  29.0  formerly smoked       1

In [6]:
data['ever_married'] = [0 if i !='Yes' else 1 for i in data['ever_married']]

In [7]:
data['gender'] = [0 if i != 'Female' else 1 for i in data['gender']]

In [8]:
data.head()

gender   age  hypertension  heart_disease  ever_married      work_type  \
0       0  67.0             0              1             1        Private   
1       0  80.0             0              1             1        Private   
2       1  49.0             0              0             1        Private   
3       1  79.0             1              0             1  Self-employed   
4       0  81.0             0              0             1        Private   

  Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
0          Urban             228.69  36.6  formerly smoked       1  
1          Rural             105.92  32.5     never smoked       1  
2          Urban             171.23  34.4           smokes       1  
3          Rural             174.12  24.0     never smoked       1  
4          Urban             186.21  29.0  formerly smoked       1

In [9]:
data = pd.get_dummies(data, columns = ['work_type', 'Residence_type', 'smoking_status'])

In [10]:
data.sample(10)

gender   age  hypertension  heart_disease  ever_married  \
343        1  39.0             0              0             1   
2337       1  38.0             0              0             1   
4633       1  50.0             0              0             1   
583        1  70.0             0              0             0   
1027       1  36.0             0              0             1   
4845       1  73.0             0              0             1   
978        0  67.0             1              1             1   
916        0  39.0             0              0             1   
679        1  58.0             1              0             1   
3294       1  23.0             0              0             0   

      avg_glucose_level   bmi  stroke  work_type_Govt_job  work_type_Private  \
343               83.51  26.4       0                   0                  1   
2337              84.31  25.9       0                   0                  1   
4633             154.67  33.8       0                   0                  1   
583               98.42  36.4       0                   0                  0   
1027              76.93  21.6       0                   0                  1   
4845              79.69  28.6       0                   0                  0   
978              254.63  31.0       0                   0                  1   
916               72.49  44.9       0                   1                  0   
679               85.83  44.0       0                   0                  1   
3294              64.94  18.8       0                   0                  1   

      work_type_Self-employed  work_type_children  Residence_type_Rural  \
343                         0                   0                     1   
2337                        0                   0                     1   
4633                        0                   0                     0   
583                         1                   0                     0   
1027                        0                   0                     0   
4845                        1                   0                     1   
978                         0                   0                     1   
916                         0                   0                     1   
679                         0                   0                     0   
3294                        0                   0                     0   

      Residence_type_Urban  smoking_status_Unknown  \
343                      0                       0   
2337                     0                       0   
4633                     1                       0   
583                      1                       0   
1027                     1                       0   
4845                     0                       0   
978                      0                       0   
916                      0                       0   
679                      1                       0   
3294                     1                       0   

      smoking_status_formerly smoked  smoking_status_never smoked  \
343                                0                            1   
2337                               0                            0   
4633                               0                            1   
583                                1                            0   
1027                               0                            1   
4845                               1                            0   
978                                0                            1   
916                                1                            0   
679                                1                            0   
3294                               0                            1   

      smoking_status_smokes  
343                       0  
2337                      1  
4633                      0  
583                       0  
1027                      0  
4845                      0  
978               

In [11]:
X = data.drop(['stroke'],axis = 1)
y = data['stroke']

In [12]:
y.value_counts()

0    4733
1     248
Name: stroke, dtype: int64

In [28]:
SMOTE = SMOTE()

X, y = SMOTE.fit_resample(X, y)

In [29]:
y.value_counts()

1    4733
0    4733
Name: stroke, dtype: int64

## Train Test Split

## Decision Tree

In [30]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=.2, random_state=42)

dt = DecisionTreeClassifier(criterion='gini', max_depth=5)
dt.fit(X_train, y_train)
dt.score(X_train, y_train), dt.score(X_test, y_test)

(0.8522187004754358, 0.839493136219641)

## Bagging

In [31]:
from sklearn.ensemble import BaggingClassifier

dt = DecisionTreeClassifier(criterion='gini', max_depth=5)
clf = BaggingClassifier(dt, n_estimators=10, random_state=42)
clf.fit(X_train, y_train)
clf.score(X_train, y_train), clf.score(X_test, y_test)

(0.8981774960380349, 0.8954593453009504)

## Random Forest

In [32]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(criterion='gini', max_depth=5, random_state=42)
rf.fit(X_train, y_train)
rf.score(X_train, y_train), rf.score(X_test, y_test)

(0.895932382461701, 0.8885955649419218)

## AdaBoost

In [33]:
from sklearn.ensemble import AdaBoostClassifier
dt = DecisionTreeClassifier(criterion='gini', max_depth=5, random_state=42)
adb = AdaBoostClassifier(dt, n_estimators=10, random_state=42)
adb.fit(X_train, y_train)
adb.score(X_train, y_train), adb.score(X_test, y_test)

(0.975567881669308, 0.9498416050686378)

## Gradient Boost

In [34]:
from sklearn.ensemble import GradientBoostingClassifier
gb = GradientBoostingClassifier(n_estimators=10, random_state=42)
gb.fit(X_train, y_train)
gb.score(X_train, y_train), gb.score(X_test, y_test)

(0.8365029054410987, 0.8352692713833157)

## XGB

In [35]:
import xgboost as xgb
xgb = xgb.XGBClassifier(n_estimators=10)
xgb.fit(X_train, y_train)
xgb.score(X_train, y_train), xgb.score(X_test, y_test)

(0.9655309033280507, 0.9588173178458289)

## Grid Search CV

In [36]:
from sklearn.model_selection import GridSearchCV

X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=.2, random_state=42)
dt = DecisionTreeClassifier(criterion='gini', max_depth=5)
params = {
    'criterion' : ['gini', 'entropy'],
    'max_depth' : [3,5,7]
}

clf = GridSearchCV(dt, param_grid=params, cv=3)
clf.fit(X_train, y_train)

clf.score(X_train, y_train), clf.score(X_test, y_test)


(0.8967247754886424, 0.8738120380147836)

## Stacking

In [37]:
from sklearn.ensemble import StackingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

clf1 = KNeighborsClassifier()
clf2 = GaussianNB()
clf3 = SVC()


estimators = [
    ('clf1', KNeighborsClassifier()),
    ('clf2', GaussianNB()),
    ('clf3', SVC())
]

sc = StackingClassifier(estimators=estimators, final_estimator=LogisticRegression())
sc.fit(X_train, y_train)
sc.score(X_train, y_train), sc.score(X_test, y_test)

(0.9536450079239303, 0.9239704329461457)

## Summary

Perbandingan akurasi terbaik dari data latih dan data uji untuk dataset Brain Stroke diperoleh dari XGBoosting dengan: <br>
**Akurasi Data Train: 0.9655309033280507** <br>
**Akurasi Data Test:  0.9588173178458289**